Fixes to the data and text string methods, updated changelog and tests for base

marty-larocque opened the pull request:
This PR addresses a number of issues uncovered in an example usecase for matplotcheck:
- Updates the changelog to address issue #130 
- Revamps all string_contains assertions in base module. This addresses issue #132 
    - This effectively resolves issue #109 by creating a new system and syntax for string_contains functions
- Fixes `assert_xydata()` incorrectly failing on some floating point numbers (issue #133)
- Rewrites allows `assert_xlabel_ydata`. The function now:
    - Evaluates wether x-labels are text strings, numeric strings, numbers, or something else
    - Intelligently decides which comparison to use based on type of x-labels
    - Addresses issue #134 
- Rewrites `get_caption()` and the functions that call it so that `get_caption()` returns a string. This addresses issue #135

marty-larocque committed:
Adding functionality for lessons

marty-larocque committed:
Added string helper to assert_caption_contains, changes to string helper.

marty-larocque committed:
More improvements to string search, additional testing.

marty-larocque committed:
Merge branch 'master' into updates_for_lessons

marty-larocque committed:
Some more changes and starting vignette

marty-larocque committed:
Merge branch 'master' into updates_for_lessons

marty-larocque committed:
Moved Vignette to another branch. Final changes for string_contains evaluation.

marty-larocque committed:
Updated Changelog

marty-larocque committed:
Had to re-rewrite assert_xlabel_ydata(). This should fix errors and improve test coverage.

marty-larocque committed:
Rewrote

marty-larocque committed:
Rewrote assert_xlabel_ydata to better accommodate numeric and string labels. Wrote tests for new functionality.

marty-larocque committed:
More test coverage improvements

lwasser commented on matplotcheck/tests/test_base_data.py:
```diff
@@ -72,6 +83,143 @@ def test_assert_xydata_xlabel_fails(pt_bar_plt, pd_df):
+
+
+def test_assert_xydata_xlabel_numeric_fails():
+    "Tests the xlabels flag on xydata correctly fails with numeric expected x-labels."
```
These data objects should be fixtures - especially considering they are used multiple times in this module. why not create a fixture that returns either both data frames or one fixture for each data frame needed? this test will then be much cleaner as will the above.  The fixture can live in this module.

lwasser commented on matplotcheck/tests/test_base_data.py:
```diff
@@ -72,6 +83,143 @@ def test_assert_xydata_xlabel_fails(pt_bar_plt, pd_df):
+    correct_df = pd.DataFrame(correct_data)
+    plot_df = pd.DataFrame(plot_data)
+
+    fig, ax = plt.subplots()
```
ok so looking at this, i think you hsould create 2 fixtues.

1. returns the plottester object for the plot_data dataframe and 
2. one that returns the dataframe used to test that the data are correct and to produce the plot. 

the first fixture above would actually required #2 above the data.frame used to test the data here.

lwasser commented on matplotcheck/tests/test_base_data.py:
```diff
@@ -72,6 +83,143 @@ def test_assert_xydata_xlabel_fails(pt_bar_plt, pd_df):
+
+
+def test_assert_xydata_xlabel_numeric():
+    "Tests the xlabels flag on xydata works with numeric expected x-labels."
```
the fixtures i mentioned below can be used here then too.

lwasser commented on matplotcheck/tests/test_base_data.py:
```diff
@@ -72,6 +83,143 @@ def test_assert_xydata_xlabel_fails(pt_bar_plt, pd_df):
+
+def test_assert_xydata_xlabel_text_fails():
+    "Tests the xlabels flag on xydata fails when testing labels with wrong text data"
+    correct_data = {
```
this dataframe should definitely be a fixture. then you can create a plot tester object from it as a second fixture. similar to below.

lwasser commented on matplotcheck/tests/test_base_data.py:
```diff
@@ -72,6 +83,143 @@ def test_assert_xydata_xlabel_fails(pt_bar_plt, pd_df):
 
+def test_assert_xydata_xlabel_text():
+    "Tests the xlabels flag on xydata works to test labels with text data"
+    data = {
```
same here -- use fixtures rather than recreating objects over and over. the tests will be greatly simplified then!

lwasser commented on matplotcheck/tests/test_base_data.py:
```diff
@@ -56,6 +58,15 @@ def test_assert_xydata_changed_data_points_only(pt_scatter_plt, pd_df):
 
 
+def test_assert_xydata_floatingpoint_error(pt_scatter_plt, pd_df):
+    for i in range(len(pd_df["A"])):
```
please add docstrings for all tests.

lwasser commented on matplotcheck/tests/test_base_data.py:
```diff
@@ -72,6 +83,143 @@ def test_assert_xydata_xlabel_fails(pt_bar_plt, pd_df):
+    """Tests that the xlabels flag on xydata correctly fails with wrong numeric
+    y-data"""
+    "Tests the xlabels flag on xydata correctly fails with numeric expected x-labels."
+    correct_data = {
```
Fixtures please!

lwasser commented on matplotcheck/tests/test_base_data.py:
```diff
@@ -72,6 +83,143 @@ def test_assert_xydata_xlabel_fails(pt_bar_plt, pd_df):
+def test_assert_xydata_xlabel_numeric_expected_string_actual():
+    """Tests the xlabels flag on xydata correctly fails with numeric expected
+    x-labels and non-numeric actial x-labels"""
+    correct_data = {
```
fixture here too!

lwasser reviewed (changes requested):
hey @ryla5068 !! this is a great start to the PR. i've left some comments. the biggest issue now is that there is a lot of repeated code that creates data for the tests.

i suggest in each case creating fixtures

1. for the data frame or data.frames needed and then 
2. for the plot tester object that ou want to create. that object can use the data frame fixture as an input to create the plot and then plottester object.

this will make the tests much cleaner!

also please be sure to add docstrings for all tests!!

lwasser commented on matplotcheck/tests/test_base_data.py:
```diff
@@ -1,7 +1,9 @@
 import matplotlib.pyplot as plt
 import numpy as np
+import pandas as pd
 import random
```
@ryla5068 is the random package actually used in these modules? pycharm is suggesting that is it not.

lwasser commented on matplotcheck/tests/test_base_data.py:
```diff
@@ -56,6 +58,15 @@ def test_assert_xydata_changed_data_points_only(pt_scatter_plt, pd_df):
+
+
 def test_assert_xydata_xlabel(pt_bar_plt, pd_df):
     "Tests the xlabels flag on xydata"
```
Docstrings should have 3 `"""` rather than 1. please fix this throughout! i noticed it several times.

lwasser commented on matplotcheck/tests/test_base_data.py:
```diff
@@ -72,6 +83,143 @@ def test_assert_xydata_xlabel_fails(pt_bar_plt, pd_df):
+        "months": ["Jan", "Feb", "Mar", "Apr", "May", "June", "July"],
+        "data": [0.635, 0.795, 1.655, 3.085, 2.64, 1.44, 1.02],
+    }
+    df = pd.DataFrame(data)
```
let's consider more expressive names for objects. this might be `expected_data_df` or `exp_data_df`

marty-larocque committed:
Added fixtures to test_base_data.py and fixed docstrings.

lwasser commented on matplotcheck/tests/test_base_data.py:
```diff
@@ -68,159 +112,116 @@ def test_assert_xydata_floatingpoint_error(pt_scatter_plt, pd_df):
-    }
+    x-labels and non-numeric actual x-labels"""
+
     plot_data = {
```
looks like this could be a fixture too... just add the new values to the fixture that you call in this test that are needed to produce the desired plot_data. it might make sense to make this plot tests object a fixture too even tho it's only used once. it just makes things a bit cleaner.

marty-larocque commented:
@lwasser I believe I fixed everything. I fixed some docstringsm, added new ones where they didn't exist, and added two (four, technically) fixtures in `test_base_data.py`. however, it looks like Travis is failing because of black.

lwasser commented on matplotcheck/tests/test_base.py:
```diff
@@ -49,3 +49,73 @@ def test_options(pt_line_plt):
+
+
+def test_assert_string_contains(pt_line_plt):
+    "Tests that assert_string_contains passes with correct expected strings."
```
All of these docstrings (for this test and the ones below) should have three `"""` quotes around them.

marty-larocque committed:
Fix docstrings

lwasser commented:
ok @ryla5068 this all looks good.i'm working on a travis fix right now over in earthpy. if it works there i'll implement it here. if it fixes things i'll move it over here as well. this PR is ready to merge otherwise. i'll ping you when you need to update this PR from master

lwasser commented:
ok @ryla5068 please fix the conflicts here and update from master. Everything is passign Ci now. and i was able to fix the `test_base_data.py` black issues as well.

marty-larocque committed:
Attempting to fix black hooks

marty-larocque committed:
Merge branch 'master' into updates_for_lessons

lwasser commented:
@ryla5068 i just merged this!! so all you need to do is update your other PR from master for the vignette and i'll look at it tomorrow. so happy CI is happy now!
